### DISCUSSION TITLE
Correct way to publish extensions with native modules?

### DISCUSSION THREAD

**Discussion by zeluspudding** [microsoft/vscode-discussions#768] [Extension Development QnA] [answered]
TLDR: What is the best practice for publishing / distributing extensions that use native / compiled modules? Forcing end users to have to rebuild modules against electron is a terrible user experience. What's the best way to correctly install such modules in an automated fashion?

Outside of developing VS Code extensions, `npm install`ing compiled modules (like `better-sqlite3`, `tree-sitter`, and `becrypt` ) works as expected since modules install themselves for the OS's default Node version. That's fine and dandy but then we try to install / use the same module within the context of a VS Code extension and we [get an error like](https://stackoverflow.com/q/71527006/3491991):

> Activating extension 'extension name' failed: The module '.../node_modules/tree-sitter/build/Release/tree_sitter_runtime_binding.node' was compiled against a different Node.js version using NODE_MODULE_VERSION 93. This version of Node.js requires NODE_MODULE_VERSION 89. Please try re-compiling or re-installing the module (for instance, using npm rebuild or npm install)..

[The reason](https://stackoverflow.com/a/71594836/3491991) for this error is that, as a big surprise to extension developers, our module dependencies are installed against the OS's Node Version instead of the version of Node built into Electron that VS Code runs on. As a result, plugin developers as well as extension end users have to rebuild the modules against the version of Node built into their current installation of VS Code's Electron, sometimes [by having to engage in 2 or 3 steps](https://stackoverflow.com/a/52796884/3491991) including deleting the `node modules` folder and installing and running the `electron-rebuild` module. 

I'm currently making an extension that suffers from these compilation miss-targettings and have spent several hours (unsuccessfully) trying to automate an exception handler that gracefully catches these issues and rebuilds modules against the VS Code's internal Node version. I.e. in trying to require `better-sqlite3` I'm doing something like this:

```
const { execSync } = require('child_process');
const electronNodeVersion = process.versions.node;
const electronModuleVersion = process.versions.modules

const Database = require('better-sqlite3');

var db
try {
  db = new Database(file_path);
} catch (error) {
  // Error occurred or better-sqlite3 is not found, so install it
  console.log("Installing better-sqlite3...");
  try {
    execSync('npm install better-sqlite3');
    execSync(`C:\\path\\to\\project\\node_modules\\.bin\\electron-rebuild.cmd -v ${electronNodeVersion}`);
    console.log("better-sqlite3 installed successfully.");
    db = new Database(file_path);
  } catch (installError) {
    console.error(`Error: ${installError}`);
  }
}
```
This looks and feels wrong (do ignore the fact I've hardcoded a module path) ... I should probably move such logic into a post installation script in package.json. All that said, such an approach would still require the `electron-rebuild` module as an extension dependency. How do I go about automating this in the package.json so that everything just works when my users install my extension? This seems to be such a common problem but I can't seem to find standard best practices.

Alternatively - are my concerns even valid? Does the act of bundling and publishing an extension to the Extension marketplace create an artifact that can be run on any vs code installation?.. I'm thinking no, but curious if anyone can confirm

**Comment by zeluspudding** [comment]
To contextualize how prevalent this problem is:
- When testing an extension with mocha, I have to rebuild modules to NODE_MODULE_VERSION 110
- When testing the same extension via the `F5` key press in VS Code, I have to rebuild modules to NODE_MODULE_VERSION 103

Makes me wonder if I'd have to rebuild modules to yet another NODE_MODULE_VERSION if I were to install the same extension but directly from the vs marketplace (my dist version).

**Comment by r3m0t** [comment]
The Node.js version is shown in Help > About. Can you use `nvm` to install the same version and use it for `npm i`?

**Comment by r3m0t** [comment]
Does your dependency use `node-gyp`? It supports compiling for a different Node version. https://www.npmjs.com/package/node-gyp

**Reply by zeluspudding** [reply]
That's the issue - end users would have to install node-gyp, which - although itself not a problem - requires having Python (and in some recommendations I've even seen C++) installed. Wouldn't be an issue if this was a dependency tree I'd internally have to setup once / when upgrading things but to force such requirements on my end users is just asking for low adoption.

**Comment by tamuratak** [comment, marked as answer]
Not officially supported. https://github.com/microsoft/vscode/issues/658

**Comment by zeluspudding** [comment]
I get the sense that generally solving this problem for native modules will involve recompiled them to WASM. In my particular case, there is a fledgling node library called [node-sqlite3-wasm](https://github.com/tndrle/node-sqlite3-wasm) that would seem to be a good alternative to `better-sqlite3`. We'll see
